### ISSUE TITLE
ENH: Have Geoseries methods called on a GeoDataFrame return a GeoDataFrame?

### ISSUE THREAD

**Issue by theroggy** [geopandas/geopandas#3434]
#### Describe the solution you'd like

When a `GeoSeries` function is called on a `GeoDataFrame`, a `GeoSeries` of the result of running the method on the geometry column is returned.

I think it would be more practical and intuitif if a copy of the `GeoDataframe` with the `geometry` column replaced with the result of the function would be returned.

This is rather a question on whether there is an explicit reason of why the behaviour is as it is or if it is rather historical or ? Because... the non-backwards-compatibility impact would most likely be large to existing code bases, and the gains are possibly/probably not worth the trouble.

#### Additional context

``` python
import geopandas as gpd
from shapely import Polygon

# Input
poly1 = Polygon([(1, 1), (8, 1), (9, 9), (1, 8)])
poly2 = Polygon([(2, 1), (8, 1), (8, 8), (2, 8)])
poly3 = Polygon([(1, 2), (9, 2), (7, 9), (2, 8)])
gdf = gpd.GeoDataFrame(geometry=[poly1, poly2, poly3])

repr_point_gdf = gdf.representative_point()
print(type(repr_point_gdf))

repr_point_series = gdf.geometry.representative_point()
print(type(repr_point_series))
```

Output:
```
<class 'geopandas.geoseries.GeoSeries'>
<class 'geopandas.geoseries.GeoSeries'>
```

**Comment by martinfleis** [comment]
While I understand, I think we're way past the time to have this discussion. This decision predates my involvement so don't know the reasons but my guess would ne that it was part of 0.1.

**Comment by theroggy** [comment]
> While I understand, I think we're way past the time to have this discussion. This decision predates my involvement so don't know the reasons but my guess would ne that it was part of 0.1.

I agree, but out of curiosity I wonder if it was a conscious choice for some reason or just not thought about... and then too late to change.

**Comment by martinfleis** [comment]
To be fair, I am not that convinced that our current behaviour is the worse one of the two. I understand that the model you suggest is what GIS software does but I always found that super annoying that a different geometric representation of the sama data makes a copy of the entire attribute table.

**Comment by theroggy** [comment]
> To be fair, I am not that convinced that our current behaviour is the worse one of the two. I understand that the model you suggest is what GIS software does but I always found that super annoying that a different geometric representation of the sama data makes a copy of the entire attribute table.

The change (ignoring the backwards compatibility issues ;-)) would not take away the option to avoid the copy of attributes, but it would give the choice:
- `gdf.representative_point()` includes a copy of the attributes
- `gdf.geometry.representative_point()` is without the copy

**Comment by m-richards** [comment]
Agree with Martin that there's no realistic way to implement a change like this given the api breakage, especially given that there is a quite simple if not so elegant alternative; `gdf = gdf.set_geometry(gdf.representative_point())`. But I do think this API discussion might be a useful reference for future libraries. I recall this being mentioned in some of the very early issues, #120 also offers a view on this, but doesn't discuss a rationale.

**Comment by theroggy** [comment]
Now we are with three agreeing. Thanks for the reference, indeed a pity there is no rationale there, or that it wasn't changed back then.

**Comment by martinfleis** [comment]
@theroggy What would be the benefit of doing that in the first place? Keep in mind that GeoPandas _extends pandas with support of geospatial operations_, not _implements GIS workflows in Python_. Those are two distinct approaches and while they do intersect, the "default" is different in different cases. I tend to think that what we have now is more sensible default within the PyData ecosystem than mirroring GIS in returning copies of whole data frames. Keep in mind that we are working in totally different context and concepts like a GeoSeries, variable or support for multiple geometry columns is non-existent in traditional GIS. I don't think we should try to get closer to GIS.

**Comment by theroggy** [comment]
> I don't think we should try to get closer to traditional GIS

I don't have significant practical experience in using "traditional" GIS programming interfaces (I suppose you mean things like ESRI/ArcPy and pyQGIS?), so I don't have detailed knowledge on how they are designed. I only know from the few encounters I have had with them that they look terrible, so in general I certainly agree that they are not something to take as an example.

Not finding it very intuitif is mainly from basic object oriented programming principles. Methods are normally defined on the level that they apply on. Applied to the topic at hand, we have a DataFrame with multiple Series (columns):
- if you want to do something with one specific column, you get the applicable column and then do what you want to do with it
- if you want to do something with the entire DataFrame, or that applies to the DataFrame, you call a method on the DataFrame

Hence, if you call a method on column/Series, you get a column back, when you call a method on a DataFrame, typically I would expect to also get a DataFrame back rather than one column. As far as I know, pandas methods follow these principles as well.

For the case of having support for multiple geometry columns, versus traditional GIS not having that, I think the current behaviour of having "a shortcut" to run a method on the "current geometry column" is a quite explicit way to lean closer to "traditional GIS"?

I am wondering now what would be the most logical way to deal with multiple geometry columns in the case of returning all columns when calling such a function on a GeoDataFrame: apply the function to all geometry-type columns or only on the "active one".

> What would be the benefit of doing that in the first place?

For an explicit benefit: now there are 2 very similar ways to get exactly the same result. Changing the behaviour of one of those two to something else that is a common use case but now needs a bit of a cluncky way to get there would sound like a good deal to me (if we were still 2014).

**Comment by martinfleis** [comment]
I thought about this for a while and could not come with a use case when I would want to return the whole GeoDataFrame. I just don't see when would it actually be useful apart from purity of API _method on gdf returns a gdf_.

**Comment by theroggy** [comment]
> I thought about this for a while and could not come with a use case when I would want to return the whole GeoDataFrame. I just don't see when would it actually be useful apart from purity of API _method on gdf returns a gdf_.

Ah... interesting... when I use geopandas I use this quite regularly. I suppose you then typically (can) overwrite the geometry column with the result of the function instead of needing the original one still afterwards or ?

**Comment by martinfleis** [comment]
Nope, I add a new column with new geometry.

**Comment by m-richards** [comment]
I'm going to close this as there's nothing realistically actionable here. Can be reopened if we find a way to progress.
